src/dataset-builder/CohortEditor.ts

```diff
@@ -206,75 +229,105 @@ type AddCriteriaSelectorProps = {
-  return h(GroupedSelect<CriteriaOption>, {
-    styles: { container: (provided) => ({ ...provided, width: '230px', marginTop: wideMargin }) },
-    isClearable: false,
-    isSearchable: false,
-    options: [
-      {
-        label: 'Domains',
-        options: _.map(
-          (domainOption) => ({
-            value: addKindToDomainOption(domainOption),
-            label: domainOption.category,
-          }),
-          snapshotBuilderSettings.domainOptions
-        ),
-      },
-      {
-        label: 'Program Data',
-        options: _.map((programDataOption) => {
-          return {
-            value: programDataOption,
-            label: programDataOption.name,
-          };
-        }, snapshotBuilderSettings.programDataOptions),
-      },
-    ],
-    'aria-label': addCriteriaText,
-    placeholder: addCriteriaText,
-    value: null,
-    onChange: (x) => {
-      if (x !== null) {
-        if (x.value.kind === 'domain') {
-          onStateChange(domainCriteriaSelectorState.new(cohort, criteriaGroup, x.value));
-        } else {
-          const criteria = criteriaFromOption(x.value);
-          if (criteria !== undefined) {
-            updateCohort(_.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, criteria));
+  const {
+    index,
+    criteriaGroup,
+    updateCohort,
+    dataset: { id: datasetId, snapshotBuilderSettings },
+    onStateChange,
+    cohort,
+  } = props;
+  return (
+    snapshotBuilderSettings &&
+    h(GroupedSelect<CriteriaOption>, {
+      styles: { container: (provided) => ({ ...provided, width: '230px', marginTop: wideMargin }) },
+      isClearable: false,
+      isSearchable: false,
+      options: [
+        {
+          label: 'Domains',
+          options: _.map(
+            (domainOption) => ({
+              value: addKindToDomainOption(domainOption),
+              label: domainOption.category,
+            }),
+            snapshotBuilderSettings.domainOptions
+          ),
+        },
+        {
+          label: 'Program Data',
+          options: _.map((programDataOption) => {
+            return {
+              value: programDataOption,
+              label: programDataOption.name,
+            };
+          }, snapshotBuilderSettings.programDataOptions),
+        },
+      ],
+      'aria-label': addCriteriaText,
+      placeholder: addCriteriaText,
+      value: null,
+      onChange: async (x) => {
+        if (x !== null) {
+          if (x.value.kind === 'domain') {
+            onStateChange(domainCriteriaSelectorState.new(cohort, criteriaGroup, x.value));
+          } else {
+            const criteriaIndex = getNextCriteriaIndex();
+            updateCohort(
+              _.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, {
+                loading: true,
+                index: criteriaIndex,
+              })
+            );
+            const loadedCriteria = await criteriaFromOption(datasetId, criteriaIndex, x.value);
+
```

Comment:
I'm struggling a bit with reviewing this because it departs significantly from the react style I'm used to. Like the _.set() call above seems like a good candidate for react state and the rest a good candidate for a use of react hooks (like useEffect). But, I think that would be a re-working of this file that's far out of scope for this PR.

Reply:
It's not at all apparent by looking at this code, but state is being used here in that `updateCohort()` is implemented by calling a react API (the result of a `useState()` call).